src/frequenz/sdk/actor/power_distributing/_component_status.py

```diff
@@ -0,0 +1,109 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Classes to track the status of components in the microgrid."""
+
+
+import enum
+import typing
+from abc import ABC, abstractmethod
+from collections import abc
+from dataclasses import dataclass
+
+from frequenz.channels import Receiver, Sender
+
+
+@dataclass
+class ComponentPoolStatus:
+    """Status of all components of a certain category in the microgrid."""
+
+    working: set[int]
+    """Set of working component ids."""
+
+    uncertain: set[int]
+    """Set of components to be used only when there are none known to be working."""
+
+    def get_working_components(self, components: abc.Set[int]) -> set[int]:
+        """From the given set of components return the working ones.
+
+        Args:
+            components: Set of components.
+
+        Returns:
+            Subset with working components.
+        """
+        working = self.working.intersection(components)
+        if len(working) > 0:
+            return working
+        return self.uncertain.intersection(components)
+
+
+class ComponentStatusEnum(enum.Enum):
+    """Enum for component status."""
+
+    NOT_WORKING = 0
+    """Component is not working and should not be used."""
+
+    UNCERTAIN = 1
+    """Component should work, although the last request to it failed.
+
+    It is blocked for few seconds and it is not recommended to use it unless it is
+    necessary.
+    """
+
+    WORKING = 2
+    """Component is working"""
+
+
+@dataclass(frozen=True)
+class ComponentStatus:
+    """Status of a single component."""
+
+    component_id: int
+    """Component ID."""
+
+    value: ComponentStatusEnum
+    """Component status."""
+
+
+@dataclass(frozen=True, kw_only=True)
+class SetPowerResult:
+    """Lists of components for which the last set power command succeeded or failed."""
+
+    succeeded: typing.Iterable[int]
+    """Component IDs for which the last set power command succeeded."""
+
+    failed: typing.Iterable[int]
+    """Component IDs for which the last set power command failed."""
+
+
+class ComponentStatusTracker(ABC):
```

Comment:
This one too could be a `BackgroundService`.

Now I wonder if we should maybe have a purely abstract `BackgroundService` class (or even protocol), and then a `BaseBackgroundService` to use as a convenience default implementation, as in here you are only defining an interface. I think `BackgroundService` should also be only an interface, as we have background services that are managing other background services instead of tasks directly, and I ended up mixing up things a bit for the sake of convenience (interface/protocol vs base mixin implementation).

But in any case for now `BackgroundService` could be used too, it will just add an unnecessary `_tasks` member to the class.

Reply:
Like this class is new, right?

Reply:
nope, it is just an abc for an existing `BatteryStatusTracker` class.

But I can look into this again as I start working on the EVChargerStatusTracker, which will be new.  And migrate this too at that point.